blog/2019-02-07-KIEServerOptaPlannerTaskAssignment.adoc

```diff
@@ -0,0 +1,215 @@
+
+NOTE: In order to build the KJAR, the project's pom.xml needs to have a `<packagin>kjar</packaging>` entry and contains
+`kie-maven-plugin`.
+
+=== Problem Description
+
+Assign each task to a spot in an employee’s queue. Each task has a duration which is affected by the employee’s affinity
+level with the task’s customer.
+
+Constraints:
+
+- [Hard] Skill: Each task requires one or more skill. The employee must possess all these skills.
+- [Soft 0] Critical tasks: Complete critical tasks first, sooner than major and minor tasks.
+- [Soft 1] Minimize makespan: Reduce the time to complete all tasks.
+- [Soft 2] Major tasks: Complete major tasks as soon as possible, sooner than minor tasks.
+- [Soft 3] Minor tasks: Complete minor tasks as soon as possible.
+
+=== Domain Model
+
+image::taskAssigningClassDiagram.png[link="taskAssigningClassDiagram.png"]
+You can read more on how to
+https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#modelAPlanningProblem[model a planning problem].
+
+=== https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#realTimePlanning[Real-time Planning]
+
+As new tasks come in and others start being implemented, the optimal solution might change. OptaPlanner handles such scenarios using
+https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#problemFactChange[ProblemFactChange].
+
+For example, if an employee starts working on a task, you can tell OptaPlanner to *pin* this task by sending a `PinTaskProblemFactChange`.
+Each `ProblemFactChange` type is implemented as a class that implements the interface `ProblemFactChange<Solution_>`.
+In the Task Assignment example, this class can be something like:
+[source,java]
+----
+public class PinTaskProblemFactChange implements ProblemFactChange<TaskAssigningSolution> {
+    ...
+
+    @Override
+    public void doChange(ScoreDirector<TaskAssigningSolution> scoreDirector) {
+        ...
+                scoreDirector.beforeProblemPropertyChanged(toBePinnedTask);
+                toBePinnedTask.setPinned(true);
+                scoreDirector.afterProblemPropertyChanged(toBePinnedTask);
+        ...
+
+        scoreDirector.triggerVariableListeners();
+    }
+}
+----
+The full class can be found https://github.com/kiegroup/kie-server-optaplanner-task-assignment/blob/master/task-assignment-kjar/src/main/java/org/kie/server/examples/optaplanner/taskassignment/kjar/domain/realtime/PinTaskProblemFactChange.java[here].
+
+===  https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#solverConfiguration[Solver Configuration]
+
+The solver configuration file determines how the solving process works. For a KJAR deployed to a KIE Server and since we are using
+Drools for score calculation, you need to use a `ksessionName`. This tells the KieContainer where to find the DRL file.
+Add this to `META-INF/kmodule.xml` file:
```

Comment:
"Add this to ..." => "Add an optataskKsession to ..."
The "this" says nothing. The code below is detailed and concrete.

What do we do VS how do we do it (we need both):
Add an optataskKsession to ... // What do we do
<kbase name=... // How do we do it